W4A1
[SumOfNumBers]
Viết chương trình Python sử dụng một trong 2 loại vòng lặp for hoặc while để
tính tống các số từ 1 đến n (n<= 1000)

In [8]:
n = int(input("Nhap n (n <= 1000): "))
if n <= 1000:
    tong = 0
    for i in range (1, n+1):
        tong += i
    print(f"tong cac so tu 1 den {n} la: {tong}")
else:
    print("vui long nhap n khong qua 1000.")

vui long nhap n khong qua 1000.


W4A2
[PrimeNumber]
VIết chương trình nhập vào một số tới khi được một số nguyên dương thì thôi.
Kiểm tra số này có phải là số nguyên tố hay không? 


In [ ]:
n =-10
while n<=0:
    n=int(input())
lst = list(range(2, 1000, 2))
print(lst)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420, 42

In [10]:
if (n==1):
    print("1 la hop so")
else:
    for i in range(2, int(n**0.5)+1):
        if(n % i ==0):
            print("n la hop so")
            break
        else:
            print("n la so nguyen to")




n la so nguyen to


Bài 1: In các số dương lẻ bé hơn 100
Bài 2: In các số dương chẵn chia hết cho 3 bé hơn 100
Bài 3: Nhập vào số nguyên dương a, in ra bảng cửu chương của a

In [5]:
n = int(input("nhap vao so duong le be hon 100 la: "))
for i in range (1, 100, 2):
    print(i, end=" ")

1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 51 53 55 57 59 61 63 65 67 69 71 73 75 77 79 81 83 85 87 89 91 93 95 97 99 

In [6]:
N = int(input("nhap cac so duong chan chia het cho 3 be hon 100: "))
for i in range (6, 100, 6):
    print(i, end=" ")

6 12 18 24 30 36 42 48 54 60 66 72 78 84 90 96 

In [17]:
while True:
    a = int(input("nhap vao so duong a: "))
    if a > 0:
        break
    print("vui long nhap so a lon hon 0")
print(f" Bang cu chuong {a}: ")
for i in range(1, 11):
    print(f"{a}x{i} = {a*i}")
    






 Bang cu chuong 5: 
5x1 = 5
5x2 = 10
5x3 = 15
5x4 = 20
5x5 = 25
5x6 = 30
5x7 = 35
5x8 = 40
5x9 = 45
5x10 = 50


Bài 4: Viết chương trình in ra hình tam giác cân có độ cao h được nhập từ bàn phím
Ví dụ: Ta nhập h = 4: 
Bài 5: Nhập vào số nguyên dương n. Tính S = 1 + 2 + 3 + 4 + … + n
Bài 6: Nhập vào số nguyên dương a, in ra tất cả các ước dương của a
Bài 7: Nhập vào số nguyên dương a, đếm số ước dương của a
Bài 8: Nhập vào số nguyên dương a và b, in toàn bộ ước dương chung của a và b
Bài 9: Nhập vào số nguyên dương a, kiểm tra xem a có phải là số chính phương hay
không? Số chính phương là số có căn bậc hai là một số nguyên VD: 1, 4, 9, 16, 25….
Bài 10: Nhập vào số nguyên dương a, kiểm tra xem a có phải là số hoàn hảo hay không.
Số hoàn hảo là số có tổng các ước (ước nhỏ hơn nó) bằng chính nó, VD: số 6, 28…
Bài 11: Nhập vào số nguyên a, nếu nhập số âm thì yêu cầu nhập lại cho đến khi người
dùng nhập đúng số dương. Nếu người dùng nhập đúng số dương thì in ra “Bạn nhập
đúng quy tắc” và dừng chương trình.
Bài 12: Nhập số nguyên dương n, Cho S(k) = 1 + 2 + 3 + … + k. Tìm k sao cho S(k)
lớn nhất nhưng nhỏ hơn n 


In [15]:
h = int(input("do cao h cua hinh tam giac: "))
for i in range (1, 1+h):
    for j in range (1, 2*h):
        if j == h - i + 1 or j == h + i - 1 or i == h:
            print("*", end="")
        else:
            print(" ", end="")
    print()

   *   
  * *  
 *   * 
*******


In [17]:
n = int(input("nhap so nguyen n: "))
print(f"S= {sum(range(1, n+1))}")

S= 10


In [19]:
a = int(input("nhap so duong a: "))
uoc = [i for i in range (1, a+1) if a%i == 0  ]
print(f"cac uoc duong cua {a}: {uoc}")
print(f"so uoc duong cua {a}: {len(uoc)}")

cac uoc duong cua 21: [1, 3, 7, 21]
so uoc duong cua 21: 4


In [ ]:
a = int(input("nhap so duong a: "))
b = int(input("nhap so duong b: "))
uocchung = [i for i in range(1, min(a,b) + 1) if a%i==0 and b%i==0]
print(f"uoc chung cua {a} và {b}: { uocchung }")

uoc chung cua 3 và 9: [1, 3]


In [23]:
a = int(input("nhap so duong a: "))
if math.sqrt(a) ** 2 ==a:
    print(f"{a} la so chinh phuong")
else:
    print(f"{a} khong phai la so chinh phuong")

NameError: name 'math' is not defined

In [24]:
a = int(input("nhap so duong a: "))
tonguoc = sum([i for i in range(1, a) if a % i==0])
if tonguoc == a:
    print(f"{a} chinh la so hoa hao")
else:
    print(f"{a} khong phai la so hoan hao")

10 khong phai la so hoan hao


In [26]:
while True:
    n = int(input("nhap so duong n: "))
    if n > 0:
        print("Ban nhap dung quy tac")
        break


Ban nhap dung quy tac


Bài 14: Nhập vào một dãy số nguyên, dừng nhập khi người dùng nhập -1. Sau khi nhập
xong, in số lớn nhất, số nhỏ nhất trong những số vừa nhập. KHÔNG ĐƯỢC dùng mảng
Bài 15: Nhập vào số nguyên dương n, đếm xem n có bao nhiêu chữ số. Làm bằng nhiều
cách nhất có thể, ko dùng mảng.
Bài 16: Nhập vào số nguyên dương n, đếm xem n có bao nhiêu chữ số chẵn, bao nhiêu
chữ số lẻ
Bài 17: Nhập vào số nguyên dương n, tính tổng các chữ số của n
Bài 18: Nhập vào một số nguyên dương n, kiểm tra xem n có phải là số dạng 3^k hay
không
Bài 19: Dãy số fibonacci là dãy số được định nghĩa như sau: 1, 1, 2, 3, 5, 8, 13,... với số
kế tiếp sẽ bằng tổng hai số trước đó. Nhập vào số nguyên dương A, hãy tìm số trong
dãy số fibonacci lớn nhất nhưng không vượt quá A
Bài 20: Nhập vào số nguyên dương n, đếm xem n có bao nhiêu chữ số là số nguyên tố. 

In [6]:
while True:
    n = int(input("Nhập số(nhập -1 để dừng): "))
    if n == -1:
        break
max_val = None
min_val = None
if max_val is None or n >max_val:
    max_val = n
if min_val is None or n< min_val:
    min_val = n
if max_val is not None:
    print(f"Số lớn nhất: {max_val}, Số nhỏ nhất: {min_val}")


ValueError: invalid literal for int() with base 10: '012345-1'

In [24]:
# 18
n = int(input("Nhap n: "))
so_ban_dau = n
dang_3k = False
if n > 0:
    while n%3==0:
        n//=3
    if n==1:
        dang_3k = True
if dang_3k:
    print(f"Số {so_ban_dau} co dang 3 mu k")
else:
    print(f"Số {so_ban_dau} khong co dang 3 mu k")

Số 9 co dang 3 mu k


W4A3
Viết chương trình tính giai thừa của một số nguyên n cho trước (n!=1×2×…×n) 


In [ ]:
n = int(input())
giai_thua= 1
if n<0:
    print("khong tinh giai thua cho so am")
else:
    n == 0
    print("0 != 1")
else:
    for i in range(1, 1+n):
    


W4A4
.[CountingDigits]
Viết chương trình nhập vào một số nguyên n và in ra màn hình số chữ số (trừ dấu)
của số đó. Không dùng kiểu dữ liệu string.

In [29]:
n = int(input())
n_goc = n
cnt = 0
if (n<0):
    n=-n
if n == 0:
    cnt = 1
else:
    while n >0:
        n = n//10
        cnt += 1
print(f" So chu so cua {n_goc} (khong dau am) la: {cnt}")

 So chu so cua 4 (khong dau am) la: 1


W4A5 [MeaningOfLife]
Khi Douglas Adam viết cuốn sách, The Hitchhiker’s Guide to the Galaxy, ông ta đã
thêm vào đó một vài lời nói đùa, và theo năm tháng nó đã trở nên nổi tiếng hơn cả
bản thân cuốn tiểu thuyết: “Câu trả lời cho câu hỏi lớn nhất về sự sống, về vũ trụ và
tất cả mọi thứ là 42.” Một vài người cuồng tín đã tốn hàng năm trời, với mọi cố
gắng nỗ lực để gán ghép ý nghĩa, biểu trưng cho con số 42 và các “sự cố” liên quan
đến nó.
Với mong muốn hái ra tiền từ sư “ám ảnh” điên rồ này, một cuốn sách mới vừa
được xuất bản, 42: Douglas Adams’ Amazingly Accurate Answer to Life, the
Universe and Everything. Cuốn sách được phát hành trùng với kỉ niệm 10 năm ngày
mất của Adams.
“Câu trả lời rất đơn giản”, như Adams đã nói, “Chỉ là một trò đùa. Tôi cần một con
số, một con số be bé, một con số tầm thường, và tôi đã chọn nó. Hai chữ số, bội của
13, một con số hoàn toàn vô nghĩa đối với mấy tay thầy chùa. Tôi ngồi trên bàn làm
việc, ngắm khu vườn và nghĩ rằng số 42 sẽ được việc lắm đây. Tôi gõ nó ra. Và
hết.”
Xuyên suốt lịch sử, các con số được gán cho những ý nghĩa đặc biệt nào đó. Plato
gọi việc nghiên cứu các biểu tượng số học” là level cao nhất của tri thức. Trong khi
Pythagoras tin rằng các con số đều có linh hồn cũng như sức mạnh kì diệu.
Trong khi đó, hàng triệu người hâm hộ Hitchhiker cho đến ngày nay vẫn kiên trì
giải mã những thứ họ cho rằng đó là những động cơ bí mật của Adams.
Hãy viết một chương trình nhận vào một số nguyên dương n và một dãy gồm n số
nguyên. Nếu n số nguyên đó có số 42 thì in ra màn hình dòng chữ "I′ve found the
meaning of life!", ngược lại in ra dòng chữ "It′s a joke!".


In [32]:
n = int(input())
for _ in range(n):
    x= int(input())
    if x == 42:
        print("I've found the meaning of life!")
        break
    else:
        print("It's a joke!")

It's a joke!
It's a joke!
I've found the meaning of life!


W4A6
"[SumPrimesInRange]
Nhập hai số nguyên dương a, b (a ≤ b). Hãy tính tổng các số nguyên tố trong đoạn
[a, b]. Mỗi dòng là một bộ test (a b)."

In [ ]:
while (True):
    a, b= map(int, input(). split())
    if a>0 and b>0 and b>a:
        break
tong=0
a=max(2, a)
for x in rang(2, int(x**0.5)+1):
    if x % i == 0:
        break
    else:
        tong = tong + x
print(tong)

W4A7
"[LargestPrimeFactor]
Nhập vào số nguyên dương n (n ≥ 2). Hãy in ra ước số nguyên tố lớn nhất của n." 

In [33]:
n = int(input("Nhập vào số nguyên dương n (n >= 2): "))
tam = n
uoc_lon_nhat = -1
while n % 2 == 0:
    uoc_lon_nhat = 2
    n //= 2
i = 3
while i * i <= n:
    while n % i == 0:
        uoc_lon_nhat = i
        n //= i
    i += 2
if n > 1:
    uoc_lon_nhat = n
print(f"Ước số nguyên tố lớn nhất của {tam} là: {uoc_lon_nhat}")

Ước số nguyên tố lớn nhất của 9 là: 3


W4A8
ReverseAddPalindrome]
Nhập số nguyên dương n. Lặp lại phép toán n = n + reverse(n) cho đến khi n là số 


W4A9
[PerfectSquaresDistinctDigits]
Nhập vào số nguyên dương n. In ra tất cả các số chính phương ≤ n có các chữ số
đôi một khác nhau (in theo thứ tự tăng dần, cùng hàng, cách nhau bởi một dấu
cách). Nếu không có, in ra (no number).

W4A10
"[LongestCollatzUpToN]
Nhập vào số nguyên dương n. Tìm số x trong [1, n] có độ dài dãy Collatz lớn nhất
và in ra x cùng độ dài dãy tương ứng (x L). Nếu có nhiều số cùng độ dài, chọn số
nhỏ nhất."

W4A11
[EvenDivisors]
Viết chương trình đếm số lượng ước số chẵn của số nguyên dương n với n < 10^6.

W4A12
[Interest]
Một người có tài khoản tiết kiệm ở ngân hàng và gửi vào X đồng với lãi suất là
0.7% mỗi tháng. Viết chương trình tính số tiền sau N tháng người ấy rút được (cả
gốc và lãi, bỏ qua phần lẻ thập phân). 

W4A13
[ClosePairNumber]
Hai số nguyên dương tạo thành một cặp số thân thiết khi chúng tuân theo quy luật:
Số này bằng tổng tất cả các ước của số kia (trừ chính nó) và ngược lại.
Ví dụ 220 và 284 là một cặp số thân thiết
220 ngoài chính nó ra, nó còn có 11 ước số khác là: 1, 2, 4, 5, 10, 11, 20, 44, 55,
110. Tổng của 11 ước này cho chúng ta 284
284 ngoài bản thân nó, còn có 5 ước số khác là: 1, 2, 4, 71, 142. Tổng của chúng
vừa đúng bằng 220
Nhập hai số từ bàn phím và kiểm tra xem chúng có phải là cặp số thân thiết không.
Nếu có, in ra màn hình "true", ngược lại, in ra "false" 

W4A14
[GCD]
Cho hai số m và n là hai số nguyên dương nhập từ bàn phím. Viết chương trình
tìm ước chung lớn nhất của hai số m

W4A15
[ChickenAndDog]
"Vừa gà vừa chó,
Bó lại cho tròn,
Ba mươi sáu con,
Một trăm chân chẵn"
Từ bài toán dân gian trên, mở rộng thành chương trình nhận đầu vào là tổng số
con và tổng số chân của gà và chó. Nếu tìm được số phù hợp, in ra số lượng gà và
số lượng chó. Ngược lại, in ra "invalid".